--- /usr/share/verilator/include/verilated_fst_c.cpp	2026-03-09 03:56:38.000000000 +0800
+++ verilated_fst_c.cpp	2026-03-14 15:54:45.209719626 +0800
@@ -31,11 +31,7 @@
 #define FST_WRITER_PARALLEL
 #define LZ4_DISABLE_DEPRECATE_WARNINGS

-// Include the GTKWave implementation directly
-#define FST_CONFIG_INCLUDE "fst_config.h"
-#include "gtkwave/fastlz.c"
-#include "gtkwave/fstapi.c"
-#include "gtkwave/lz4.c"
+#include "gtkwave/fstapi.h"

 #include <algorithm>
 #include <iterator>
@@ -81,6 +77,8 @@
     m_fst = fstWriterCreate(filename, 1);
     fstWriterSetPackType(m_fst, FST_WR_PT_LZ4);
     fstWriterSetTimescaleFromString(m_fst, timeResStr().c_str());  // lintok-begin-on-ref
+    fstWriterSetDate(m_fst, "Sat Jan 01 01:23:45 2000\x0a");
+    fstWriterSetVersion(m_fst, "Verilator");
     if (m_useFstWriterThread) fstWriterSetParallelMode(m_fst, 1);
     constDump(true);  // First dump must contain the const signals
     fullDump(true);  // First dump must be full for fst
@@ -332,67 +329,49 @@
 void VerilatedFstBuffer::emitEvent(uint32_t code) {
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], "1");
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], 1, 1);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitBit(uint32_t code, CData newval) {
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], newval ? "1" : "0");
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], 1, newval);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitCData(uint32_t code, CData newval, int bits) {
-    char buf[VL_BYTESIZE];
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
-    cvtCDataToStr(buf, newval << (VL_BYTESIZE - bits));
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], buf);
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], bits, newval);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitSData(uint32_t code, SData newval, int bits) {
-    char buf[VL_SHORTSIZE];
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
-    cvtSDataToStr(buf, newval << (VL_SHORTSIZE - bits));
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], buf);
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], bits, newval);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitIData(uint32_t code, IData newval, int bits) {
-    char buf[VL_IDATASIZE];
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
-    cvtIDataToStr(buf, newval << (VL_IDATASIZE - bits));
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], buf);
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], bits, newval);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitQData(uint32_t code, QData newval, int bits) {
-    char buf[VL_QUADSIZE];
     VL_DEBUG_IFDEF(assert(m_symbolp[code]););
-    cvtQDataToStr(buf, newval << (VL_QUADSIZE - bits));
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], buf);
+    fstWriterEmitValueChange64(m_fst, m_symbolp[code], bits, newval);
 }

 VL_ATTR_ALWINLINE
 void VerilatedFstBuffer::emitWData(uint32_t code, const WData* newvalp, int bits) {
     int words = VL_WORDS_I(bits);
-    char* wp = m_strbufp;
-    // Convert the most significant word
-    const int bitsInMSW = VL_BITBIT_E(bits) ? VL_BITBIT_E(bits) : VL_EDATASIZE;
-    cvtEDataToStr(wp, newvalp[--words] << (VL_EDATASIZE - bitsInMSW));
-    wp += bitsInMSW;
-    // Convert the remaining words
-    while (words > 0) {
-        cvtEDataToStr(wp, newvalp[--words]);
-        wp += VL_EDATASIZE;
-    }
     m_owner.emitTimeChangeMaybe();
-    fstWriterEmitValueChange(m_fst, m_symbolp[code], m_strbufp);
+    fstWriterEmitValueChangeVec32(m_fst, m_symbolp[code], bits, newvalp);
 }

 VL_ATTR_ALWINLINE
